overlord/snapstate/autorefresh_gating.go

```diff
@@ -0,0 +1,223 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/overlord/ifacestate/ifacerepo"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+var gateAutoRefreshHookName = "gate-auto-refresh"
+
+type affectedSnapInfo struct {
+	Restart        bool
+	Base           bool
+	AffectingSnaps map[string]bool
+}
+
+func affectedByRefresh(st *state.State, updates []*snap.Info) (map[string]*affectedSnapInfo, error) {
+	all, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	var bootBase string
+	if !release.OnClassic {
+		deviceCtx, err := DeviceCtx(st, nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get device context: %v", err)
+		}
+		bootBaseInfo, err := BootBaseInfo(st, deviceCtx)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get boot base info: %v", err)
+		}
+		bootBase = bootBaseInfo.InstanceName()
+	}
+
+	byBase := make(map[string][]string)
+	for name, snapSt := range all {
+		if !snapSt.Active {
+			delete(all, name)
+			continue
+		}
+		inf, err := snapSt.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		// optimize: do not consider snaps that don't have gate-auto-refresh hook.
```

Comment:
(nitpick^2): maybe `optimization`, the current "optimize" sounds (to me) like something we should do not that it's done here.